File: python/day1.py
#!/usr/bin/env python3

from pathlib import Path
import typing as t

DIAL_SIZE = 100


def process(input_path: Path, start_pos: int = 50) -> t.Tuple[int, int]:
    pos = start_pos
    stop_count = pass_count = 0

    with open(input_path) as f:
        for line in f:
            line = line.strip()
            sign = +1 if line[0] == 'R' else -1
            value = int(line[1:])

            # number of movements to reach first zero
            offset = DIAL_SIZE - pos if sign * pos < 0 else pos
            pass_count += (offset + value) // DIAL_SIZE

            pos += sign * value
            pos %= DIAL_SIZE

            if pos == 0:
                stop_count += 1

            #print(f"{line}, pos: {pos}, pass_count: {pass_count} stop_count: {stop_count}")

    return (stop_count, pass_count)


def main():
    input_folder = Path(__file__).absolute().parent.parent / 'input'
    input_path = input_folder / 'day1.txt'

    (stop_count, pass_count) = process(input_path)

    print(f"Stopped at 0 {stop_count} times")
    print(f"Passed 0 {pass_count} times")

if __name__ == '__main__':
    main()

File: rust/src/day1.rs
use core::fmt;
use std::{fs::File, io::{BufRead, BufReader}};

use super::{verbosity, input_dir};

static DIAL_SIZE: i64 = 100;

fn try_parse_line<'a>(line: &'a str) -> Option<(i64, i64)> {
    if line.len() < 2 { return None; }
    Some((
        match line.chars().next().unwrap().to_ascii_uppercase() {
            'L' => -1,
            'R' =>  1,
            _ => return None,
        },
        (&line[1..]).parse().ok()?,
    ))
}

pub fn process<I, E>(lines: I, start_pos: i64, verbosity: u8) -> Result<(i64, i64), String>
where I: IntoIterator<Item = Result<String, E>>,
      E: fmt::Display
{
    let mut pos = start_pos;
    let mut stop_count = 0;
    let mut pass_count = 0;

    for line in lines.into_iter() {
        let line = line.map_err(|e| format!("Error reading file: {}", e))?;
        let (sign, value) = try_parse_line(&line).ok_or_else(|| format!("Invalid line {:?}", line))?;

        let offset = if sign * pos < 0 { DIAL_SIZE - pos } else { pos };

        pass_count += (offset + value) / DIAL_SIZE;
        pos += sign * value;
        pos = pos.rem_euclid(DIAL_SIZE);

        if pos == 0 { stop_count += 1 }

        if verbosity > 0 {
            println!("{:<4} pos: {:>2}, passes: {:>4}, stops: {:>4}", line, pos, pass_count, stop_count);
        }
    }

    Ok((stop_count, pass_count))
}


pub fn run(test: bool) -> Result<(), String> {
    let verbosity = verbosity();

    let mut input_path = input_dir().to_owned();
    input_path.push(if test { "day1_test.txt" } else { "day1.txt" });

    println!("input_path: {:?}", input_path);

    let file = File::open(input_path).expect("Failed to open input file");
    let (stop_count, pass_count) = process(BufReader::new(file).lines(), 50, verbosity)?;

    println!("Stopped at 0 {} times", stop_count);
    println!("Passed 0 {} times", pass_count);

    Ok(())
}